bot/rest/server.py

```diff
@@ -36,14 +38,25 @@ def stop(self):
 
 def create_http_server(bot, port=8080):
     app = HTTPServer(bot, port)
+    flask_cors.CORS(app)
 
+    register_controllers(app, bot)
+    register_error_handlers(flask=app)
+    return app
+
+
+def register_controllers(app, bot):
     controller = [
         HealthController(),
         GuildController(bot),
         ResetRosterController(bot),
         OtherController(bot),
     ]
-
     for ctrl in controller:
         app.register_blueprint(ctrl.api)
-    return app
+
+
+def register_error_handlers(flask: Flask):
+    @flask.errorhandler(HTTPException)
+    def _handle_error(e: HTTPException):
```

Comment:
this smells **a lot** like creating control flow through the use of exceptions, which becomes visible in `_createGuild` of the `guild_controller`. The werkzeug module seems to facilitate this kind of control flow by offering exceptions for specific HTTP-codes and global exception handlers, but beside that, it is considered an anti-pattern.

Reply:
This error handler just ensures, that any response will be a valid Json, so clients do not have to deal with different content types.

Reply:
That is of course a fine addition, but in `bot/rest/controller/guild_controller.py` we have the exact situation where either a `jsonify`'d response is returned, or a `BadRequest`-error is raised. This makes the path the code takes more opaque, since it may bypass a caller in case of the thrown error while just bubbling all the way up to the set error-handler. This is not an issue in our case _yet_, but an anti-pattern nonetheless.

Reply:
Hm, the used "abort(int,str)" function basically does the same, even sais it in the docs: 
`Raises an :py:exc: HTTPException for the given status code or WSGI
    application.`

Reply:
Absolutely! It is recommended behaviour by the authors of the library; this is an issue in how the library itself encourages the handling, which I think is bad style in general. As I said, this is probably not a problem for our usecase yet, but I always see exceptions as `goto`s in disguise and I am therefore very sceptical when they are used to create control flow which basically could circumvent cleanup-code in callers or later segments of the function the exception is thrown in. Similar goes for `break`, `continue`, and `return` in the middle of a function, btw.

I would have stuck with returning a code and a message to the caller and have them create the appropriate HTTP response, but I am also not too eager on changing it back, if you are convinced by the current implementation.

Reply:
I can see the issue since Python does not have checked Exceptions. On the other hand this makes return values of functions much more complicates, as basically every function would have to return some kind of error-situation response next to it's result.

Python itself seems to be using Exceptions even in "expected" flow: https://jeffknupp.com/blog/2013/02/06/write-cleaner-python-use-exceptions/

Exceptions are **meant to be handled** by the caller, as different callers might have different ways of dealing with a specific error, or even errors in general.
A webserver will return an error page, while a command line application will exit with some specific error code.
In the case of flask, you just raise a BadRequest Exception and let the error handler deal with it.
This also prevents lots of dublicated code, as you don't need to create a bunch of ErrorResponses in each controller.

To summarize: I agree on Exceptions beeing antipattern as beeing used for the expected control flow (e.g. "NoResultsFoundException" -> use None or an emtpy list as return) **but** for "Dead Ends" in the application, you do not want to deal with passing the error upwards with each method call in the stack on your own. That's what exceptions are meant to do.

Reply:
I agree on the "dead ends" situation, but I don't believe this is a dead end as failing when trying to create a guild is kind of an expected outcome, not a critical error. But I think in the end this may be an issue with my personal reservations against unstructured programming, which exceptions are inherently a part of, so I may be biased and the approach you suggested is probably fine as is.